tests/suite/test_transport_server_tcp_load_balance.py

```diff
@@ -0,0 +1,175 @@
+        """
+        Function to revert a TransportServer resource to a valid state.
+        """
+        patch_src = f"{TEST_DATA}/transport-server-status/standard/transport-server.yaml"
+        patch_ts(
+            kube_apis.custom_objects,
+            transport_server_setup.name,
+            patch_src,
+            transport_server_setup.namespace,
+        )
+
+    def test_number_of_replicas(
+        self, kube_apis, crd_ingress_controller, transport_server_setup, ingress_controller_prerequisites
+    ):
+        """
+        The load balancing of TCP should result in 4 servers to match the 4 replicas of a service.
+        """
+        original = scale_deployment(kube_apis.apps_v1_api, "tcp-service", transport_server_setup.namespace, 4)
+        wait_before_test()
+
+        result_conf = get_ts_nginx_template_conf(
+            kube_apis.v1,
+            transport_server_setup.namespace,
+            transport_server_setup.name,
+            transport_server_setup.ingress_pod_name,
+            ingress_controller_prerequisites.namespace
+        )
+
+        pattern = 'server .*;'
+        num_servers = len(re.findall(pattern, result_conf))
+
+        assert num_servers is 4
+
+        scale_deployment(kube_apis.apps_v1_api, "tcp-service", transport_server_setup.namespace, original)
+        wait_before_test()
+
+    def test_tcp_request_load_balanced(
+            self, kube_apis, crd_ingress_controller, transport_server_setup, ingress_controller_prerequisites
+    ):
+        """
+        Requests to the load balanced TCP service should result in responses from 3 different endpoints.
+        """
+        port = transport_server_setup.public_endpoint.tcp_server_port
+        host = transport_server_setup.public_endpoint.public_ip
+
+        print(f"sending tcp requests to: {host}:{port}")
+
+        endpoints = {}
+        for i in range(20):
+            client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            client.connect((host, port))
+            response = client.recv(4096)
+            endpoint = response.decode()
+            print(f' req number {i}; response: {endpoint}')
+            if endpoint not in endpoints:
+                endpoints[endpoint] = 1
+            else:
+                endpoints[endpoint] = endpoints[endpoint] + 1
+            client.close()
+
+        assert len(endpoints) is 3
+
+        result_conf = get_ts_nginx_template_conf(
+            kube_apis.v1,
+            transport_server_setup.namespace,
+            transport_server_setup.name,
+            transport_server_setup.ingress_pod_name,
+            ingress_controller_prerequisites.namespace
+        )
+
+        pattern = 'server .*;'
+        servers = re.findall(pattern, result_conf)
+        for key in endpoints.keys():
+            found = False
+            for server in servers:
+                if key in server:
+                    found = True
+            assert found
```

Comment:
no way to assert a string?

Reply:
can we add a test where two TS have same listener port and the one deployed later causes port collision?

Reply:
I've added an additional test to cover port collision and resolution.